### PR TITLE
Async evaluation

### PR DESCRIPTION
**Pull request by YangFei1990**
1.Add numba jit compile to accelerate the mask paste part in evaluation
2.Running eval on all nodes instead of the master node
3.Add support for async evaluation and early stop once the target is reached
4.Use a custom coco evaluation from Nvidia [repo](https://github.com/NVIDIA/cocoapi/)
5.Add ompi_bind script to use the cpu cores more efficient

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tensorpack/libinfo.py [hunk 1]
```diff
@@ -63,4 +63,4 @@
 # These lines will be programatically read/write by setup.py
 # Don't touch them.
 __version__ = '0.9.0.1'
-__git_version__ = __version__
+__git_version__ = "v0.0.0-87-g3ac448a-dirty"
```
Comment (pidajay) [inline tensorpack/libinfo.py]: Is this intentional?

File: MaskRCNN/eval.py [hunk 2]
```diff
@@ -378,31 +405,21 @@ def _eval(self):
         if cfg.TRAINER == 'replicated':
             all_results = multithread_predict_dataflow(self.dataflows, self.predictors)
         else:
-            filenames = [os.path.join(
-                logdir, 'outputs{}-part{}.json'.format(self.global_step, rank)
-            ) for rank in range(hvd.local_size())]
-
-            if self._horovod_run_eval:
```
[the remark this answers is not in the document]
Comment (YangFei1990) [reply MaskRCNN/eval.py]: > By removing this line are you assuming all eval will be horovod based? Can this run on single GPU?

This flag is used to set eval only in master node (i.e. hvd.rank() == hvd.local_rank()). It has nothing to do with horovod actually. Check line 350 where the flag is removed.
Also we can run on single GPU with/without horovod.

File: MaskRCNN/eval.py [hunk 3]
```diff
@@ -413,3 +430,97 @@ def _trigger_epoch(self):
...
+            else:
+                self.dataflow = get_eval_dataflow(self._eval_dataset,
+                                              shard=hvd.rank(), num_shards=hvd.size())
+
+
+    def _build_predictor(self, idx):
+        return self.trainer.get_predictor(self._in_names, self._out_names, device=idx)
+
+    def _before_train(self):
+        if hvd.rank() == 0:
```
[the remark this answers is not in the document]
Comment (YangFei1990) [reply MaskRCNN/eval.py]: > Will horovod be initialized when running on single GPU?

Horovod will be initialized if the Horovod trainer is selected.

File: MaskRCNN/train.py [hunk 4]
```diff
@@ -309,8 +412,9 @@ def check_and_log(cmd):
             starting_epoch=cfg.TRAIN.STARTING_EPOCH
         )
 
-
-        if is_horovod:
+        if args.async_eval:
+            trainer = AsyncTrainer(average=True)
```
[the remark this answers is not in the document]
Comment (YangFei1990) [reply MaskRCNN/train.py]: > This is a nit. Maybe it would be clearer to call this 'AsyncHorovodTrainer'?

Good point. I changed it. Check my [new commit](https://github.com/aws-samples/mask-rcnn-tensorflow/pull/11/commits/77261306d7ce332fb838abf68d1e21c1a4495590)

File: tensorpack/libinfo.py (hunk 1 above)
Comment (YangFei1990) [reply tensorpack/libinfo.py]: > Is this intentional?

This might be introduced from patch. I changed it back

**Comment by johnbensnyder** [comment]
Ran on 4 nodes p3.16 last night. Everything seems to work fine.

File: MaskRCNN/eval.py [hunk 5]
```diff
@@ -59,7 +60,7 @@ def _scale_box(box, scale):
     scaled_box[3] = y_c + h_half
     return scaled_box
 
-
+@jit
```
Comment (johnbensnyder) [inline MaskRCNN/eval.py]: Not necessary for this PR, but we should add in explicit types for the jit, that should give us a slight performance boost and get rid of most of the warnings.
